### ISSUE TITLE
Major refactoring of storage classes.

### ISSUE THREAD

**Issue by ytsmiling** [optuna/optuna#1170]
## Motivation

Current storage class implementations have several issues listed in the following.
- Inconsistent error handlings between backends.
- Significant bottlenecks in the storage class on training with / querying to RDB.
- O(trial^2) running time even with random-sampler and CMA-ES.
- Undefined consistency models for multi-worker optimizations.

Some of these limitations come from the optuna's origin, that is, optuna was initially
developed for deep-learning applications with ~1000 trials.
Given increasing applications optuna, it will be a good time to reconsider the design
and implementations of storage classes.

This issue is meant to be used for open discussions and PR tracking.

## Roadmaps

Some of the roadmaps come from committers' discussions, but they are open for discussions. Comments are appreciated.

### Documentation and implementation consistency

- [ ] Document consistency models for storage classes.
- [ ] Document each method, especially concerning error-handling.
- [ ] Add tests and fix implementation on error-handling.
- [ ] Log when new studies are created in all backends.
- [ ] Define the default study name.

### Performance improvements

- [ ] Add cache for unfinished trials.
- [ ] Support bulk updates of trial attributes.
- [ ] Support bulk read of DB contents, and make storages sync only once per trial.
- [ ] Support incremental load of DB contents.
- [ ] Support indexing of finished trials.

### Refactoring

- [ ] Remove unnecessary return values.
- [ ] Add more default implementations in `BaseStorage` class.
- [ ] Removed unnecessary backend-specific method implementations.

### Miscellaneous issues

- [ ] Support heart-beat of trials.
- [ ] Support multi-studies in in-memory storage.
- [ ] Fix race-condition in waiting queue.
- [ ] Nan-handling in best-trial determination.
- [ ] Raise errors when a single trial is modified from multiple workers at the same time.
- [ ] Remove multi-threading support and add multi-processing support.

**Comment by ytsmiling** [comment]
I'd like to summarize recent changes.

## Major Changes
One of the major changes is that trials are only loaded only when `storage.read_from_remote_storage` in RDB backend. This solved a bug that trial looks different set of trials under multi-worker training (#1298). Additionally, reduced communication cost improved overall performance. Another large change is that we add multi-worker tests, which has long been missing (#1497).

## Bulk update
Bulk-update is still in progress When `_CachedStroage` class is used, trials are updated using the `_update_trial` method in `RDBStorage`, but the method does not perform bulk-update. Additionally, we immediately write `user_attrs`, `system_attrs`, and `intermediate_values` into storages, and this requirement degrades performance. Committer discussions reached an agreement that using other workers' `intermediate_values` and `system_attrs` in pruners are more important than improving optimization speed, and thus the requirement exists.

## O(1) suggests in CMA-ES
We removed unnecessary deep-copies and the runtime significantly decreased. However, the running time still takes O(trial) time per parameter suggestion for CMA-ES, NSGA-2, and others. Additionally, we don't have a cache for samplers and samplers. For example, TPE needs to recalculate the sorted list every time they suggest parameters, which is a significant bottleneck of the current TPE implementation. We may support the indexing of finished trials or add some caching mechanisms for samplers to solve these issues.

**Comment by hvy** [comment]
Let me leave a note that details are missing for several items marked contribution-welcome. If developers would be interested in tackling storage related refactoring (given the information already provided here), it'd be a good idea to first clarify the details here.

**Comment by ytsmiling** [comment]
I think many details have changed from when this issue was created. Additionally, some of the issues marked as contribution-welcome actually require a lot of discussions. For example, "Remove multi-threading support and add multi-processing support" has several ways to address. Also, concerning "Add more default implementations in BaseStorage class." and "Removed unnecessary backend-specific method implementations.", I'm not sure whether we should keep adding implementations to `BaseStorage` or alternatively wrap storage classes using `_CachedStorage`, which will need discussion.
Shall we discuss how to tackle the above (or possibly whether we should address the issues or not) before we ask for contributions?

**Comment by hvy** [comment]
As you say, I'm imagining we have settle each of those discussion when tackling those controversial items. At the same time, some items are still more straight forward, such as the ones under "Documentation and implementation consistency"? Maybe we should clarify this difference in the description?

**Comment by ytsmiling** [comment]
I agree that we should clarify the differences. Before tackling some of the issues, I think we first need to reorganize tasks and reconsider a roadmap for the future backend implementations. Let's have a chat to discuss them.

**Comment by hvy** [comment]
Separated out 

> (🎉 contribution-welcome🎉) Log when new studies are created in all backends.

into https://github.com/optuna/optuna/issues/1622.
